Refactor/dev more cleanup plus unnamed semaphores

drbitboy opened the pull request:
~This PR/MR should not be merged until after PR/MR #39 is merged.~

Ready to go!

drbitboy committed:
Convert named semphores in /dev/shm/sem.* to unnamed semaphores in shmim

Fix sizeof style

DasVinch commented:
Thanks @sanfords for putting it together.
There seems to now be a conflict due to merging #39 into dev.

I'm also stalling a little bit, I'd like to find time to grep-patrol through cacao and milk and check for breaking changes (and how many of those...) in pre-ISIO-API code.

drbitboy commented:
rebased this branch, conflicts have been resolved

drbitboy commented:
> I'm also stalling a little bit, I'd like to find time to grep-patrol through cacao and milk and check for breaking changes (and how many of those...) in pre-ISIO-API code.

Take your time!

DasVinch commented:
See also https://github.com/milk-org/milk/pull/10 and https://github.com/cacao-org/cacao/pull/23 that port compliance to these changes in extensions.

drbitboy commented:
> See also [milk-org/milk#10](https://github.com/milk-org/milk/pull/10) and [cacao-org/cacao#23](https://github.com/cacao-org/cacao/pull/23) that port compliance to these changes in extensions.

we could/should also eliminate all direct references, in milk and cacao, to ...{IMAGE}.semptr[semindex] that use system sem_* routines, and replace them with to ImageStreamIO_sem* interface routines.

drbitboy committed:
Add missing [map] pointer update in ImageStreamIO.c when opening existing shmim (ImageStreamIO_read_sharedmem_image_toIMAGE)

drbitboy commented:
Any movement on this?

drbitboy committed:
Merge branch 'dev' into refactor/dev-more-cleanup-plus-unnamed-semaphores

drbitboy committed:
Create new internal interfaces to calculate shmim sizes and addresses

drbitboy committed:
Clean up some issues with the previous commit

drbitboy committed:
Have ImageStreamIO_createIm_gpu use the ImageStreamIO_image_sizing_from_scratch interface to size shmim instead of doing the sizing internally

drbitboy committed:
Use sizing in read...Image(...) routine

drbitboy committed:
ImageStreamIO.c

- Ensure created shmim file is a new empty file
  - Unlink existing file if needed
- Add inode utilities specific to shmim framework
- Close shmim in ImageStreamIO_closeIm
- Increase generality and robustness
- Improve comments
- Eliminate unneeded return codes
- Clean up miscellaneous items

drbitboy committed:
Fix and clean up ImageStreamIO_check_image_inode

drbitboy committed:
Add inode check to C header

drbitboy committed:
Fix typo

drbitboy committed:
Unfix typo, fix other typo

drbitboy committed:
tests/ImageStreamIO_subTest_Operations.cpp

- change operational test between two processes transferring data
  between them using semaphores for synchronization:
  Change:
  - Parent process creates a new shmim file just before posting
    (incrementing) a semaphore for which the child process is waiting.
    - New shmim creation unlinks the old shmim file and associates
      the new shmim's inode with the same pathname
  - Child process times out waiting on the semaphore in the old shmim.
  - Child process uses the ImageStreamIO_check_image_inode routine to
    confirm inode of the old shmim it is using does not match the inode
    of the new shmim with the same path name.
  - Child process opens the new shmim
  - Child process waits on and decrements the semaphore posted by the
    parent process, which allows the data transfer to continue.

ImageStreamIO.c

- Move where in the code that the inode is written to the shmim metadata

DasVinch commented on ImageStruct.h:
We should probably get rid of SEMFILEDATA altogether.

drbitboy replied:
Yes, we could do that.  I did it this way at this point to minimize the changes to the **map += sizeof(...) \* K** logic.

We could also declare **md->semptr** to be (sem_t*) instead of (sem_t**), and then **md->semptr+_N_** would be passed to the system sem_dothis routines, and then we would not have to allocate space for the **md->sem** count of pointers and assign them individually.

DasVinch commented:
Since this is structurally changing, I propose to increment version number to 2.0. @oguyon 

(Brian Carcich/drbitboy)>>>  Agreed.

DasVinch commented:
Also, I think merging #44 has caused a conflict.

DasVinch commented:
_FYI,_ Ideally, I'd like to deploy this in prod on July 12 (sorry for the long wait).

drbitboy commented:
> Also, I think merging #44 has caused a conflict.

That was due to a typo in the old code (sizeof(sem_t*) should have been sizeof(sem_t**) in the malloc size calculation), but because it was a pointer either way, it did not matter.

Also, #44 will be obsolete with the consolidation of the **map += ...** code.

drbitboy commented:
> _FYI,_ Ideally, I'd like to deploy this in prod on July 12 (sorry for the long wait).

No problem, thanks for getting into this.  I'll sort out the conflicts.

drbitboy committed:
Merge branch 'dev' into refactor/dev-more-cleanup-plus-unnamed-semaphores

drbitboy committed:
Increment IMAGESTRUCT_VERSION string to 2.00 for structural changes i.e. unnamed semaphores

drbitboy commented:
> Since this is structurally changing, I propose to increment version number to 2.0. @oguyon
> 
> (Brian Carcich/drbitboy)>>> Agreed.

Done, IMAGESTRUCT_VERSION is now "2.00"

oguyon committed:
add missing stdbool header

oguyon committed:
fix non-shared mem image creation bugs

drbitboy commented on ImageStreamIO.c:
```diff
@@ -349,20 +353,34 @@ int ImageStreamIO_typesize(
+    case _DATATYPE_COMPLEX_DOUBLE:
+        return SIZEOF_DATATYPE_COMPLEX_DOUBLE;
+    default:
+        break;
```
This difficult-to-read style is probably what caused an error in a previous commit.

Just sayin'

drbitboy commented on ImageStreamIO.c:
```diff
@@ -1097,28 +1172,70 @@ errno_t ImageStreamIO_createIm_gpu(
+        {
+            printf("Memory allocation error %s %d\n", __FILE__, __LINE__);
+            abort();
+        }
```
For shared, non-local memory, keywords are in the shmim, so this malloc should only be performed when the memory is local, so move this back into the **else** clause of **if (shared == 1) {...} else**.

drbitboy commented on ImageStreamIO.c:
```diff
@@ -1172,7 +1289,7 @@ errno_t ImageStreamIO_createIm_gpu(
     }
 
-    // initialize keywords
+// initialize keywords
```
???

drbitboy commented on ImageStreamIO.c:
```diff
@@ -1097,28 +1172,70 @@ errno_t ImageStreamIO_createIm_gpu(
         }
         image->md->shared = 0;
         image->md->inode = 0;
```
Malloc that sets image->md->kw pointer needs to be in this **else (shared!=1)** clause or else image->md->kw pointing to keywords in shmim assigned in **if (shared==1)** clause will be overwritten.

drbitboy committed:
Add keyword test to, improve comments in, and remove obsolete logging from, operations test

drbitboy commented:
I am working on this some more to address the non-shared issues that Olivier raised.  It should be done well before mid-July.

Brian Carcich

drbitboy committed:
Revert some of the changes by @oguyon

oguyon committed:
address compiler warnings

drbitboy committed:
Fix style

drbitboy committed:
Rearrange logic in shmim sizing

drbitboy committed:
ImageStreamIO.c

- Improve ImageStreamIO_image_sizing* routines
  - Move more assignments into ImageStreamIO_image_sizing* routines
  - Handle non-shared, process-local memory case
  - More robust handling of invalid arguments
- Reorder, refactor logic
- Improve comments and style

drbitboy committed:
Add a second 'forked-child' test to validate non-shared, process-local calls to ImageStreamIO_createIm_gpu(...)

drbitboy committed:
Update routine headers; return invalid arg instead of failure for invalid location passed to ImageStreamIO_createIm_gpu/_image_sizing* routrines

drbitboy committed:
Move divider comments so [git diff dev] shows a more useful result; remove trailing spaces

drbitboy committed:
Correct a comment

drbitboy committed:
Update change log

drbitboy committed:
Utilize unused wCBcycle; tighten up code; make #if*-#else-#endif's more legible

drbitboy committed:
Refactor for clarity, simplicity, robustness

ImageStreamIO.c

- Refactor semaphore routines
- Remove obsolete commented code (printf logging)

ImageStreamIO_unitTest.cpp

- Update test for changed returned error code

drbitboy committed:
Make ImagesStreamIO_closeIm and some sempahore routines more robust

drbitboy committed:
Ensure sizing calculations handle invalid arguments correctly

drbitboy committed:
Convert testing from gtest (Google test framwork) to Catch2 framework; first pass

drbitboy committed:
Add CMake changes for tests/ISIO_Catch2

drbitboy commented:
Update:  hold off on this PR until PR #47 is resolved

drbitboy committed:
Remove old tests/CMakeLists.txt

drbitboy committed:
Remove unnecessary compile options

a-sevin commented:
Could we consider to merge #49 directly into your branch ?
I don't see any issue to merge it here, it simply adds a placeholder for the cudaMalloc. This placeholder is only for the internal computations in the GPU (loopcounter, timestamp, ...) to avoid code them in the image (and therefore degrade it). 
However, as it's really a minor modification which no impact if you keep the value to 0, it could be done afterwards.

DasVinch commented:
Apologies that it took 3 months. I believe the day has finally cometh.
- I'm gonna make sure I compile jointly with https://github.com/milk-org/milk/pull/10 and https://github.com/cacao-org/cacao/pull/23.
- Then I'm gonna deploy on one of the scexao server.
- And then confirm the merge.
